feat(application-controller): Add support for rollback multi-source applications

JorTurFer opened the pull request:
Signed-off-by: Jorge Turrado <jorge.turrado@scrm.lidl>

Currently, ArgoCD support multi-source applications and that's a really nice feature, but using multi source, end users have to sacrify the option of executing fast rollbacks because the rollbacks are not supported for multi-source applications. 
This PR adds the support for rollback multi-source applications (as single source applications do)

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates? (I don't think so, does it?)
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

Fixes https://github.com/argoproj/argo-cd/issues/12580

JorTurFer commented:
Please, put special attention to the front because I'm not front-end developer and maybe I have done something terriblely wrong

JorTurFer commented:
This is how it looks in the frontend:
![image](https://github.com/argoproj/argo-cd/assets/36899226/3c4f857a-5e57-438f-b246-95cd86a7ed4a)

![image](https://github.com/argoproj/argo-cd/assets/36899226/5efaa069-4ea8-4cbe-8f4f-924264f7530f)

ownyrd commented:
Thanks for implementing this. Hope it get`s merged soon :)

crenshaw-dev commented:
Thanks for the PR! We're past feature freeze for 2.8, so I'll queue this up for the 2.9 roadmap.

JorTurFer commented:
> Thanks for the PR! We're past feature freeze for 2.8, so I'll queue this up for the 2.9 roadmap.

Sure 😄 
Do you have any ETA for reviewing the PR? No rush at all, it's just for fixing merge conflicts when you are going to review it to not be rebasing it all the time xD

JorTurFer commented:
> Makes sense! At least 3 weeks out, I'm focusing on getting 2.8 ready for GA for now.

After vacations (second half of August) looks as a nice moment to rebase the PR with latest changes 😄  
Thanks for the info 🙇

GuerreiroLeonardo commented:
we are waiting on this one!1

kewei5zhang commented:
Hoping to have this merged soon. 🙏

JorTurFer commented:
Hi @crenshaw-dev 
Is it now a good moment to rebase my PR? 😄

crenshaw-dev commented on reposerver/repository/repository.proto:
```diff
@@ -36,6 +36,8 @@ message ManifestRequest {
     // This is used to surface "source not permitted" errors for Helm repositories
     string projectName = 25;
+    // This is used during multi-source rollback to replace versions in "ref" sources
+    bool isRollback = 26;
```
I'm not sure I'm a fan of introducing the "rollback" concept to the repo-server. Can we instead just collect the sync configuration in the controller or API server and pass that to the repo-server as if we were just syncing to that commit?

JorTurFer replied:
Let me check it

JorTurFer replied:
I have removed the logic from here and I have moved it into the state

crenshaw-dev reviewed (dismissed):
Only had time for a quick glance, opened one can of worms. :-)

JorTurFer commented:
Sorry, I needed more time that I expected for rebasing the PR 🤦  (the summer break has erased my memory 100%)

JorTurFer commented:
I have removed the logic from here and I have moved it into the state

JorTurFer commented:
> Any news on this? Do you guys need an extra pair of hands?

It's waiting a review 😄

crenshaw-dev commented:
@Pehesi97 we need lots of extra pairs of hands. 😆

GuerreiroLeonardo commented:
Any predictions on when it will be released?

JorTurFer commented:
@GuerreiroLeonardo , if you could help them reviewing the code, I think this will be faster for sure 😄 
I guess that we are a bit overloaded atm with the summer break and the last release

GuerreiroLeonardo commented:
> @GuerreiroLeonardo , if you could help them reviewing the code, I think this will be faster for sure 😄 I guess that we are a bit overloaded atm with the summer break and the last release
@JorTurFer JorTurFer I wish I could, I have no knowledge on the argoCD code. I am simply a user right now hahah. I gues it would take a lot of hours until I am proficient enough to give any relevant contribution. Sorry =/

gioppoluca commented:
We also consider this PR important, do not have to capabilities to look at the code, but could test it once it gets released.

jsoref commented on util/argo/ref_sources.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+// GetRefSources creates a map of ref keys (from the sources' 'ref' fields) to information about the referenced source.
+// This function also validates the references use allowed characters and does not define the same ref key more than
```
```suggestion
// This function also validates the references use allowed characters and do not define the same ref key more than
```

gioppoluca replied:
Is a spelling error on a comment a good reason to ask for a review on such a feature? Seems a good way to make contributors stop contributing.

JorTurFer replied:
I took the code from argo.go and I moved it as it was and i didn't notice it. I'll update it 😄

jsoref commented on util/argo/ref_sources.go:
```diff
@@ -0,0 +1,62 @@
+				}
+				refKey := "$" + source.Ref
+				if _, ok := refKeys[refKey]; ok {
+					return nil, fmt.Errorf("invalid sources: multiple sources had the same `ref` key")
```
it'd be really nice if errors like this reported at least a pair of sources so that users have something to chase

JorTurFer replied:
I took the code from `argo.go` and I moved it as it is but I could add the key to the message if you think is useful

jsoref replied:
My goal is to be able to read a message and figure out how to address it. If this error will be shown as is, I'd want to see the id of the thing (is it an `application/<x>`?) and two `sources` that had the same `ref` (preferably with the `ref`). If it's going to be pasted w/ other things, then as long as once it's pasted it has those bits, that'd make it usable.

JorTurFer replied:
I'd try to improve the message 😄 
I'll review the logic trying to figure out the best way to print the message

jsoref reviewed (commented):
I'd want to see:
* design doc
* user facing documentation
* a ui demo (so I can kick the tires) -- the picture in https://github.com/argoproj/argo-cd/pull/14124#issuecomment-1604044195 should be in the description.

Note that I'm not wearing a project owner hat, this what I'd want as a drive-by reviewer / UI person / end user.

jsoref commented:
In:
![](https://user-images.githubusercontent.com/36899226/248238669-5efaa069-4ea8-4cbe-8f4f-924264f7530f.png)

Why do I see the same revision twice?

If I can't kick the tires, could you at please kick the tires for me?

User Story:
1. User has a "multi-source application" (some sample content for me to look at)
2. User deploys it (show picture of screens)
3. User makes a change (?) (show change, picture of how things change)
4. User needs to trigger a rollback (show path through UI that user takes to trigger it)
5. User sees rollback complete (show how the UI looks)
6. User wants to review to see what rollback(s) have happened (show path through UI for this).

JorTurFer commented:
> design doc
user facing documentation
a ui demo (so I can kick the tires) -- the picture in https://github.com/argoproj/argo-cd/pull/14124#issuecomment-1604044195 should be in the description.

Hi @jsoref 
I appreciate your feedback but you couldn't see those things just because the [PR template](https://github.com/argoproj/argo-cd/blob/master/.github/pull_request_template.md) doesn't require them and I'm new contributing to this project.
I followed it and I have tried to do my best, but based on the documentation I had, those things weren't required. If you point me to any doc about how can I provide the UI demo, I'd be more than happy providing it. I'd really appreciate any guide on that part because I'm totally lost in front side.

> Why do I see the same revision twice?

Because probably I used the same repository for both sources, sorry. I didn't think that it was important.

> If I can't kick the tires, could you at please kick the tires for me?

Except for how the multiple sources are presented, all the other points are exactly the same as the current scenario. It's true that I added the extra sources information to the UI (using the parenthesis), but I can remove it if you think that it's wrong.

JorTurFer commented:
@jsoref, is the rest of the code okey for you? I'd like to rebase my branch and address all the commits at once to not need to rebase the branch a lot of times

jsoref commented:
Sorry, don't worry too much about it. I can't find any real documentation/pictures for the core feature.

There's some text about it, but that's about it:
https://argo-cd.readthedocs.io/en/latest/user-guide/commands/argocd_app_rollback/
https://argo-cd.readthedocs.io/en/latest/proposals/application-name-identifier/#upgrade-downgrade-strategy

I've never used the feature (we are gitops and essentially only roll forward).

Fwiw, here's what I see in my history and rollback (it includes a ⋮ for each row, I'm not sure why I only see it for your first row):
<img width="1312" alt="image" src="https://github.com/argoproj/argo-cd/assets/2119212/bd222c6c-6b06-4aff-9953-ee1a887e923e">

If you use it, does the rollback thing actually work?
<img width="1312" alt="image" src="https://github.com/argoproj/argo-cd/assets/2119212/34489738-defb-4b25-807b-84637ed3d055">

---

The rest of the code doesn't seem problematic to me, but I'm not speaking as an owner, I'm speaking with an end user / doc hat on.

JorTurFer commented:
> Fwiw, here's what I see in my history and rollback (it includes a ⋮ for each row, I'm not sure why I only see it for your first row):

I see exactly the same, but you are seeing 2 versions from a single source application and I see a single version of a multi source application:
![image](https://github.com/argoproj/argo-cd/assets/36899226/93f7f8f5-70a0-438c-a3a8-0fdca8f4669d)

![image](https://github.com/argoproj/argo-cd/assets/36899226/2004821c-0b01-4fc1-abda-63b96580556b)

> If you use it, does the rollback thing actually work?

With my PR, I can enter the panel and inside it, I see the button and I can rollback the application using it. With current ArgoCD code, the button to enter the panel is directly disabled, so I can't see them nor click on it:
https://user-images.githubusercontent.com/22163931/220861338-4fa629d9-c7f7-45e7-91c4-5ecaf2f63506.png

> The rest of the code doesn't seem problematic to me, but I'm not speaking as an owner, I'm speaking with an end user / doc hat on.

Yeah, don't worry. It's because I have limited time and I prefer to gather all the feedback and then rebase my branch and address all the feedback at once. This will be the second time that I'll have to rebase my branch, and I prefer to do all the things at once 😄

jsoref commented:
Ok, that's useful. I'm not a fan of this UI.

the solid

---

is doing strange work.

In general, when you look at one of these things, do you care about all the sources at the same time or only one at a time?

If you're looking at a history, are you likely to want to compare `source 2` at time `A` with `source 2` at time `B`? If yes, then having tabs for sources might be better than having them all linearly.

It's possible that just adding headings for each source would be useful.

(A picture where the sources are clearly different would be helpful.)

JorTurFer commented:
> In general, when you look at one of these things, do you care about all the sources at the same time or only one at a time?

IDK because I haven't had the option tbh. I guess that it'd depend on the case... 

> If you're looking at a history, are you likely to want to compare source 2 at time A with source 2 at time B? If yes, then having tabs for sources might be better than having them all linearly.

This isn't either for a single source rollback. I mean, having current version and previous version in different tabs looks super useful to me too, but it's not the current behavior neither for the single source and redesign that UI/UX is out of this PR's scope (I'm not frontend developer and being honest, I don't think that I'm able to do it neither). e.g: A single source with 50-100 helm values is render as a panel with a long scroll, and you need to scroll down until the end, and the you have to click on the other version, expanding the selected version and collapsing all the non-selected versions.

> the solid is doing strange work.

You are totally right, but I didn't have any other idea tbh... I thought in it because it was quick and not invasive, but any suggestion about how to replace it without refactoring the whole panel is more than welcomed 😄 

> (A picture where the sources are clearly different would be helpful.)

You are right, but I don't have a live version now to take a picture, sorry. if you think that it's important, I could prepare it tomorrow and take the picture but I hope that this picture explains the main idea:
![image](https://github.com/argoproj/argo-cd/assets/36899226/2004821c-0b01-4fc1-abda-63b96580556b)

GuerreiroLeonardo commented:
> I'd want to see:
> 
> * design doc
> * user facing documentation
> * a ui demo (so I can kick the tires) -- the picture in [feat(application-controller): Add support for rollback multi-source applications #14124 (comment)](https://github.com/argoproj/argo-cd/pull/14124#issuecomment-1604044195) should be in the description.
> 
> Note that I'm not wearing a project owner hat, this what I'd want as a drive-by reviewer / UI person / end user.



Thank you for your review @jsoref

cregev commented:
Hi guys,
In which version do expect to release it?

VisorX commented:
hello world,

I may get a promotion if this feature is added soon 😄 

Thank you

crenshaw-dev commented:
> I may get a promotion if this feature is added soon 😄

100% considering adding your promotion to the release notes when this ships. 😆 

For folks wondering when this is gonna get merged, I really can't make promises right now. We're trying our best to get good things into 2.10 though.

JorTurFer commented:
> I may get a promotion if this feature is added soon 😄

Will I see your promotion somehow (💵) ? 🤣 🤣 

Now seriously, folks from ArgoCD are doing an awesome job driving the project and this PR is huge, I totally understand that they need more time than usually for reviewing and checking it and I hope you too 😄

GuerreiroLeonardo commented:
Hi guys, any updates on when it is coming online?

GuerreiroLeonardo commented:
@JorTurFer JorTurFer

GSZoominfo commented:
> > I may get a promotion if this feature is added soon 😄
> 
> 100% considering adding your promotion to the release notes when this ships. 😆
> 
> For folks wondering when this is gonna get merged, I really can't make promises right now. We're trying our best to get good things into 2.10 though.

will it be part of 2.10?

it does not seem to be part of the release candidate 
https://blog.argoproj.io/argo-cd-v2-10-release-candidate-f69ba7bf9e06

crenshaw-dev commented:
I'm afraid not. I dived into the initial review but quickly realized I wouldn't be able to get it finished by RC date.

GSZoominfo commented:
> I'm afraid not. I dived into the initial review but quickly realized I wouldn't be able to get it finished by RC date.

any estimation of when this feature will be released?

GuerreiroLeonardo commented:
Happy new year my friends! It could be happier if this feature was online hahaha

GuerreiroLeonardo commented:
@crenshaw-dev any news?

ChrisLanks commented:
Hey team, thanks for putting all this together. We really appreciate everything you guys do! Seriously.

But, we are eagerly waiting for this new feature as it will dramatically improve our developer's work streams. Is there any shot at getting this prioritized for the next ArgoCD release?

GuerreiroLeonardo commented:
> Only had time for a quick glance, opened one can of worms. :-)

Are these resolved?

GuerreiroLeonardo commented:
@ChrisLanks according to ardocd roadmap, this should be release in 03-20-2024, But I really am not hopeful since another issue are still in review as well =(
<img width="358" alt="image" src="https://github.com/argoproj/argo-cd/assets/59945594/a5383bbf-bd2c-4f92-9c79-afe94071e9b8">

ishitasequeira reviewed (commented):
@JorTurFer , I am thinking of getting the PR reviewed and prioritized. Would you be able to rebase the PR and resolve conflicts? 

Also, @keithchong has a [proposal for multiple sources](#17106 ) where he has referenced this PR. It would be good to see if the UI is aligned for multiple sources support.

JorTurFer commented:
Hello @ishitasequeira,
> @JorTurFer , I am thinking of getting the PR reviewed and prioritized. Would you be able to rebase the PR and resolve conflicts?

I can rebase my PR if you have plans for reviewing it, but please confirm it to me as it's not the first rebase I've done. I don't have so much time nowadays, so I can't guarantee that the UI is fully aligned with the proposal tbh (I'm not frontend expert and it requires more time to me).

Could we continue with the UI as it is currently? Maybe other folk could just work on the UI changes having the backend ready. Or maybe there is any other folk willing to fork my PR and starting from here 🤷

ishitasequeira commented:
@JorTurFer , I am surely planning to spend time on reviewing this PR and have added it to my TODO. We can surely continue with backend work and expose it via cli (if you want to add cli integration) as part of this PR. We can add the UI support as part of @keithchong 's proposal [proposal for multiple sources](https://github.com/argoproj/argo-cd/pull/14124#17106) and take it from there. 

WDYT @JorTurFer @keithchong ??

GuerreiroLeonardo commented:
Hello guys. @JorTurFer any updates?

JorTurFer commented:
I want to rebase the PR this week, sorry
KubeCon + easter have been a huge plan breakers xD

JorTurFer commented:
> @JorTurFer Is there a chance we'll see this feature available by the end of this week?

🤷🤷🤷🤷🤷🤷 

IDK, I'll try, but I've started the rebase and there are too many conflicts that I have to solve, and I don't have the context of the changes in conflict. Currently, this feature is not necessary for us anymore (we use full gitops with autosync), so I'm doing this on my own free time that I have to share with my own maintained project xD

GuerreiroLeonardo commented:
@JorTurFer @ishitasequeira I can believe this day is actually happening😭

JorTurFer commented:
omg, more merge conflicts 🤦

GuerreiroLeonardo commented:
@ishitasequeira Any updates on when you're gonna be able to review this?

ishitasequeira commented:
@GuerreiroLeonardo  I have it added to my current sprint.

ishitasequeira commented on controller/sync.go:
```diff
@@ -108,21 +108,24 @@ func (m *appStateManager) SyncAppState(app *v1alpha1.Application, state *v1alpha
+	if rollback {
+		// rollback case
 		if app.Spec.HasMultipleSources() {
-			sources = app.Spec.Sources
+			sources = state.Operation.Sync.Sources
 		} else {
-			// normal sync case (where source is taken from app.spec.source)
-			source = app.Spec.GetSource()
+			source = *state.Operation.Sync.Source
 			sources = make([]v1alpha1.ApplicationSource, 0)
 		}
 	} else {
-		// rollback case
+		// normal sync case (where source is taken from app.spec.sources)
 		if app.Spec.HasMultipleSources() {
-			sources = state.Operation.Sync.Sources
+			sources = app.Spec.Sources
 		} else {
-			source = *state.Operation.Sync.Source
+			// normal sync case (where source is taken from app.spec.source)
+			source = app.Spec.GetSource()
 			sources = make([]v1alpha1.ApplicationSource, 0)
 		}
 	}
```
I think, we can avoid checking whether the app is Multisource App or not on line 111. 
```suggestion
        rollback := (syncOp.Sources != nil && len(syncOp.Sources) > 0) || syncOp.Source != nil
	if !rollback {
		// normal sync case (where source is taken from app.spec.sources)
		if app.Spec.HasMultipleSources() {
			sources = app.Spec.Sources
		} else {
			// normal sync case (where source is taken from app.spec.source)
			source = app.Spec.GetSource()
		}
	} else {
		// rollback case
		if app.Spec.HasMultipleSources() {
			sources = state.Operation.Sync.Sources
		} else {
			source = *state.Operation.Sync.Source
		}
	}
	if sources == nil {
		sources = make([]v1alpha1.ApplicationSource, 0)
	}
```

Let me know what you think.

ishitasequeira commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -205,6 +205,11 @@ func (s ApplicationSources) Equals(other ApplicationSources) bool {
 }
 
+// IsZero returns true if the application source is considered empty
+func (a ApplicationSources) IsZero() bool {
+	return len(a) == 0
+}
```
Considering we need to check for Sources to be not nil before we call `IsZero` anytime, I think we can better utilize this function as below:
```suggestion
func (a ApplicationSources) IsNullOrEmpty() bool {
	return (a != nil && len(a) == 0) || a == nil
}
```

WDYT?

ishitasequeira commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -251,6 +256,11 @@ func (a *ApplicationSource) AllowsConcurrentProcessing() bool {
 	return true
 }
 
+// IsRef returns true when the application source is of type Ref
+func (a *ApplicationSource) IsRef() bool {
+	return a.Ref != ""
+}
```
Is this function being used anywhere? I could not find it to be referenced anywhere.

JorTurFer replied:
nice catch! It was used in a previous version that I did, but now it's not used

JorTurFer replied:
In any case, I have seen some places to use it and align the code with other checks like IsHelm, so I'll use it. WDYT?

ishitasequeira replied:
If we can replace `ref != nil` everywhere, it would be good to keep it.

ishitasequeira reviewed (commented):
@JorTurFer I am yet to test the changes, but do check my initial feedback.

JorTurFer commented on server/application/application.go:
```diff
@@ -1923,9 +1971,10 @@ func (s *Server) Rollback(ctx context.Context, rollbackReq *application.Applicat
 		return nil, status.Errorf(codes.InvalidArgument, "application %s does not have deployment with id %v", a.QualifiedName(), rollbackReq.GetId())
 	}
-	if deploymentInfo.Source.IsZero() {
+	if deploymentInfo.Source.IsZero() && deploymentInfo.Sources.IsZero() {
```
Can it be nil? It's just an empty slice, so checking `nil` or the len should be the same, and it doesn't panic because I'm not dereferencing it. You can check the behavior here: https://go.dev/play/p/UF45REm9OKE

ishitasequeira replied:
Thanks for the playground link. That helps understand it.

ishitasequeira commented on util/argo/ref_sources.go:
```diff
@@ -0,0 +1,62 @@
+// In case of rollback, this function also updates the targetRevision to the proper revision
+func GetRefSources(ctx context.Context, opts GetRefSourcesOptions) (argoappv1.RefTargetRevisionMapping, error) {
+	refSources := make(argoappv1.RefTargetRevisionMapping)
+	if len(opts.Sources) > 1 {
```
We can avoid the check `len(opts.Sources) > 1` as the loop would not be executed if `opts.Sources` is empty.

JorTurFer replied:
Yeah, but you still can have 1 single item in sources slides, that's the reason. Originally this func received the app itself and checked `if spec.HasMultipleSources()` and this is the closer option.

I can remove it if you still think that it's not required

ishitasequeira reviewed (commented):
While testing the feature out, it seems, the application is not able to handle rollback from `source` to `sources` field or vice versa. To reproduce, 

1. Create a single multi-source application
2. Update the application to be of a single source, i.e. remove `spec.sources` and add `spec.source`.
3. Try rolling back to the previous version. It does not rollback the app to `spec.sources`.

JorTurFer commented:
> While testing the feature out, it seems, the application is not able to handle rollback from `source` to `sources` field or vice versa. To reproduce,

yeah, I didn't support changing between both because I thought that it wasn't an scenario, but I can check and try to extend the PR to support it too, let me check it. Do the other changes work as expected?

JorTurFer commented:
@ishitasequeira , I've updated the PR supporting changes between single source and multi source. PTAL when you have some time.
I've also rebased my branch because there was a merge conflict related with a self generated file

JorTurFer commented:
I'm not sure why the [docker image is not built well](https://github.com/argoproj/argo-cd/actions/runs/8681453504/job/23804073800?pr=14124), but I'd say that it's not related with my changes due to the message:
```
ERROR: failed to solve: process "/bin/sh -c apt-get update && apt-get install --no-install-recommends -y     openssh-server     nginx     unzip     fcgiwrap     git     git-lfs     make     wget     gcc     sudo     zip &&     apt-get clean &&     rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*" did not complete successfully: exit code: 100
```

keithchong commented on ui/src/app/applications/components/application-deployment-history/application-deployment-history.tsx:
```diff
@@ -79,28 +76,83 @@ export const ApplicationDeploymentHistory = ({
-                                                }}
-                                                details={details}
-                                            />
+                                            <div className='row'>
```
I'm working on adding the multi-source support in what is currently the Parameters tab.  For example, see [PR](https://github.com/argoproj/argo-cd/pull/17589).  In order to better distinguish between the sources, I thought that adding showing the index within the sources field would help.  So around this area, what do you think about adding a title here, something like what is depicted in my screenshots.   It could be as simple as showing `Source {i+1}`

On a separate note, I made the current Edit card a collapsible section, and this is something 'new' to the UI. If there are many sources, and the list of parameters can get really long, it could get unwieldy and difficult to use. What you have here is fine as is, since the info is read-only.  In order to get your PR merged, I think we could make them these rows collapsible in a future PR (in order to align the multi-source presentation).

keithchong commented on ui/src/app/applications/components/application-operation-state/application-operation-state.tsx:
```diff
@@ -93,7 +93,15 @@ export const ApplicationOperationState: React.StatelessComponent<Props> = ({appl
+            value: (
+                <Revision
+                    repoUrl={utils.getAppDefaultSource(application).repoURL}
+                    revision={utils.getAppDefaultSyncRevision(application) + utils.getAppDefaultSyncRevisionExtra(application)}
```
Check if the extra part should be part of the revision field value, which might be included as part of the hyperlink

keithchong replied:
Just to follow up on this, the link is 'broken' because the 'and (1) more' phrase is included

<img width="654" alt="revisionLink" src="https://github.com/argoproj/argo-cd/assets/7726022/93949ab9-5562-4d1b-baf3-f0ee45c791dc">

keithchong commented on ui/src/app/applications/components/application-status-panel/application-status-panel.tsx:
```diff
@@ -32,16 +33,11 @@ const sectionLabel = (info: SectionInfo) => (
-                    <i className='fa fa-ellipsis-h' />
-                </button>
-            )}
+            {onClick && <button className='application-status-panel__more-button' />}
```
There was a 'More' button here before Ishita disabled it. Is this the new intended behavior?

keithchong commented on ui/src/app/applications/components/application-operation-state/application-operation-state.tsx:
```diff
@@ -93,7 +93,15 @@ export const ApplicationOperationState: React.StatelessComponent<Props> = ({appl
         });
     }
     if (operationState.syncResult) {
-        operationAttributes.push({title: 'REVISION', value: <Revision repoUrl={utils.getAppDefaultSource(application).repoURL} revision={operationState.syncResult.revision} />});
```
Previously, the revision value wouldn't be displayed until the operation has finished.  Seems like in this new behavior, the revision represents the most recent, successfully synced revision, and therefore will always show here, even if the operation is in progress.  Is this the intended design?  Is the if check on syncResult needed?

keithchong commented on ui/src/app/applications/components/application-status-panel/application-status-panel.tsx:
```diff
@@ -107,13 +102,14 @@ export const ApplicationStatusPanel = ({application, showDiff, showOperation, sh
                         {application.spec.syncPolicy?.automated ? 'Auto sync is enabled.' : 'Auto sync is not enabled.'}
                     </div>
-                    {application.status && application.status.sync && application.status.sync.revision && !application.spec.source.chart && (
+                    {application.status && application.status.sync && application.status.sync.revision && revision && !application.spec.source.chart && (
```
Could `application.status.sync.revision && revision` be simplified to just `revision`?

keithchong commented on ui/src/app/applications/components/application-status-panel/application-status-panel.tsx:
```diff
@@ -130,29 +126,28 @@ export const ApplicationStatusPanel = ({application, showDiff, showOperation, sh
                             {appOperationState.phase} <Timestamp date={appOperationState.finishedAt || appOperationState.startedAt} />
                         </div>
-                        {(appOperationState.syncResult && appOperationState.syncResult.revision && (
+                        {(appOperationState.syncResult && revision && (
```
See my other comment regarding the operation sync result

ishitasequeira commented:
> I'm not sure why the [docker image is not built well](https://github.com/argoproj/argo-cd/actions/runs/8681453504/job/23804073800?pr=14124), but I'd say that it's not related with my changes due to the message:
> 
> ```
> ERROR: failed to solve: process "/bin/sh -c apt-get update && apt-get install --no-install-recommends -y     openssh-server     nginx     unzip     fcgiwrap     git     git-lfs     make     wget     gcc     sudo     zip &&     apt-get clean &&     rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*" did not complete successfully: exit code: 100
> ```

Thanks @JorTurFer for updating the PR with the use-case. It seems to be working now. The issue you mentioned has been fixed in master. Once you rebase your PR, the issue should get resolved.

GuerreiroLeonardo commented:
Hi @JorTurFer any idea when you're gonna be able to looking into these issues?

JorTurFer commented:
I really appreciate the feedback @keithchong! Sadly, I don't have enough time nowadays to address it, there have been a lot of changes in the UI since I checked it time ago and I'm not frontend developer, so understand the UI requires a huge time to me. As the backend is almost ready, and it works, would be you willing to be invited to my repo and help me with those changes directly there? I feel that you have more context about how the things have to be done in front side.

keithchong commented:
Hi @JorTurFer , reverting or taking away your UI changes might be a bit extreme. I'll see if I can fix the link issue and any other 'straight-forward' issues.  For the other concerns, I think we should open follow-on Issues and address them in separate PRs, hopefully for the next release. Aligning the look and feel and adding back the 'More' button should not be blocking your PR.  IMO, from the issues I saw, the incorrect status for the sync operation in progress is probably higher priority, because it is misleading information.

JorTurFer commented:
Sorry for being totally disconnected for a long time, these last weeks have been really complicated :(

I expect to have a bit more time soon, which is the current status? I've seen that you have merged another PR related with moultisource UI. Should I rebase my PR or are you working on it too?

ishitasequeira commented:
@JorTurFer I am out this week. But I can follow up next week on the PR. Meanwhile, afaik @keithchong has taken a stab at fixing the minor issues in your PR. I will probably wait for him to share an update.

keithchong commented:
Hi @JorTurFer , I believe I've addressed the issues I pointed out in the PR review.  See the following animated gif.  My changes are based on top of your PR changes, so I believe you can go ahead and get approval to merge your PR, with the caveat that any of the outstanding UI issues I outlined will be addressed in a follow-on PR.  I still have to polish and test some more, but I should be able to create the PR shortly after your changes are merged.

![UpdatesToHistoryPR](https://github.com/argoproj/argo-cd/assets/7726022/8ccfb3c0-36a3-487e-81ba-6de6624b002e)

The video shows the commit link phrases (eg. _commitHash_ and 2 more) are consistent and the link is only on the commit hash. It shows the triple-dots button (formerly known as the More button) enabled and shows more info on each of the source commit or version, and other information, just like before for single source apps.  Also it just shows that these changes are based on your PR with the history and rollback button enabled.

Also, in the following screenshot, the sync operation failed, so clicking on the triple-dots button (formerly known as the More button), doesn't do anything, and I think this was the original behaviour):

<img width="1060" alt="SyncOperationError" src="https://github.com/argoproj/argo-cd/assets/7726022/81d4825c-9816-411f-841e-f94b06c84291">

ishitasequeira reviewed (commented):
@JorTurFer would you be able to rebase your PR? We can get the changes merged and have @keithchong raise a separate PR on top of your changes with the fixes he has identified.

ishitasequeira commented on server/application/application.go:
```diff
@@ -1498,7 +1499,67 @@ func (s *Server) RevisionMetadata(ctx context.Context, q *application.RevisionMe
+		// using the specific revisionId
+		for _, h := range a.Status.History {
+			if h.ID == versionId {
+				h := h
```
any reason for the need of re-assigning the value for this variable?

JorTurFer replied:
Yeah, as source is a pointer, I redefine the var within the scope to hide the iteration one: https://go.dev/blog/loopvar-preview

JorTurFer replied:
https://go.dev/play/p/fhnEXSoWXlb?v=goprev

ishitasequeira reviewed (approved):
LGTM!! Thanks @JorTurFer !!

JorTurFer commented:
Thanks @crenshaw-dev @ishitasequeira and @keithchong  for your support! ❤️

richardpj commented:
This is wonderful news! Thank you all for your efforts.

GuerreiroLeonardo commented:
@crenshaw-dev @ishitasequeira Thank you for your efforts! Could you please confirm on which version will this feature be released?

afreyermuth98 commented:
Hello ! 
Any idea when this feature would be available on a release ? :)

blakepettersson commented:
@afreyermuth98 this should be available in 2.12.
